### DISCUSSION TITLE
Using Sparse Tensors

### DISCUSSION THREAD

**Discussion by chinmay5** [pyg-team/pytorch_geometric#2414] [Q&A] [answered]
Thank you so much for the wonderful library. Unfortunately, I am struggling a bit with the basics. I wanted to understand how to correctly use Sparse Tensors and interchanging between different formats. 

1. Creating a `Data` object.

```
x = torch.as_tensor(x_data, dtype=torch.float)
   edge_index = torch.as_tensor( [[int(e[0]), int(e[1])] for e in zip(*final_graph.nonzero())],  dtype=torch.long)
    edge_features = [[final_graph[int(e[0])][int(e[1])]] for e in zip(*final_graph.nonzero())]
    edge_features = torch.as_tensor(np.concatenate(edge_features), dtype=torch.float)  # .unsqueeze(1)
    y_data = torch.as_tensor(y_data, dtype=torch.long).squeeze()
    data = Data(x=x, edge_index=edge_index.t().contiguous(), y=y_data, edge_attr=edge_features)
    return data
```
2. Switching to SparseTensors
```
transform =  T.ToSparseTensor()
data = transform(data=data)
inv_transform = T.ToDense()
data = inv_transform(data)
```

I did the latter part to see if I can switch between the two without many issues. However, I keep getting the error:-
```
Traceback (most recent call last):
  File "/home/chinmayp/workspace/population-gcn/geometric_dataset.py", line 62, in <module>
    data = inv_transform(data)
  File "/home/chinmayp/setup/pytorch/lib/python3.6/site-packages/torch_geometric/transforms/to_dense.py", line 16, in __call__
    assert data.edge_index is not None
AssertionError
```
I am not exactly sure how to proceed with the dataset creation and their interchange. Any help would be highly appreciated.

**Comment by wsad1** [comment, marked as answer]
`ToSparseTensor` has an argument called `remove_edge_index`. Set it to false while initializing it. When its true the function sets `data.edge_index` to None.
```
def __init__(self, remove_edge_index: bool = True, fill_cache: bool = True)
```
@rusty1s , do you think its useful to add a `remove_edge_index` argument to `ToDense`.

**Reply by chinmay5** [reply]
Thank you for the response and it works as explained. I have a follow-up though. I am new to the framework and it would be great if you can also check if the way I have assigned different attributes looks correct. I have a N x N adjacency matrix which is weighted and stored in the variable `final_graph`.

**Reply by wsad1** [reply]
Checkout out the [dense_to_sparse](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/sparse.html#dense_to_sparse) function.
With that you can do.
```
edge_index, edge_weight= dense_to_sparse(final_graph) # final_graph is a tensor.
```
